### PR TITLE
Agent v3

### PR DESCRIPTION
**Pull request by petemounce**
Support agent v3.x in hopefully a non-breaking manner (the existing template was renamed and kept verbatim, and different variables are in play for the v3 things).

### CODE REVIEW

File: tasks/main.yml
Comment (petemounce) [inline tasks/main.yml:16]: I initially had a problem with the previous syntax when I added the variable expansion to it - this version of the syntax makes it clearer, I think, to read and notice the need for quotes.

File: defaults/main.yml
Comment (hectcastro) [inline defaults/main.yml]: Can you please update the `README` with the new variables?

File: defaults/main.yml
Comment (hectcastro) [inline defaults/main.yml]: I tried testing this behavior as part of https://github.com/azavea/ansible-buildkite-agent/pull/2, but was unable to reproduce it.

File: defaults/main.yml
Comment (hectcastro) [inline defaults/main.yml:16]: Since this only gets used in `buildkite_agent_tags`, do you think we could get rid of the variable entirely and just leave  `buildkite_agent_tags` with a `queue=` default.

File: defaults/main.yml
Comment (hectcastro) [inline defaults/main.yml]: Should this default to `buildkite-agent bootstrap` based on the comment in https://github.com/azavea/ansible-buildkite-agent/pull/3/files#diff-c065fbb7e1b1961b255807b0096329f5R22?

File: defaults/main.yml
Comment (hectcastro) [inline defaults/main.yml]: Just as a heads up that I'd be OK with breaking backwards compatibility with the v2 agent. We'd likely just bump the major version of the Ansible role to help communicate that.

File: defaults/main.yml
Comment (petemounce) [reply defaults/main.yml:16]: I'd like to not do that, because of how I'm intending to roll out changes to our own buildkite cluster. 

I am planning to use a git sha1 from my config-management-and-packer repo as the value of the `queue` tag - this means that I know the provenance of the agents, and also means that when I want to make changes I can do so in a safe way because new agents get assigned to an isolated queue. Pipeline steps can then be transitioned across by changing their queue to that of the new agent cluster.

I also thought keeping it separate makes a bit of logical sense since it's a special case for a tag.

What do you think?

File: defaults/main.yml
Comment (petemounce) [reply defaults/main.yml]: I think I'll remove this entirely because of how I've (since) decided to deploy changes to the agent cluster (see https://github.com/azavea/ansible-buildkite-agent/pull/3#discussion_r181770528).

File: defaults/main.yml
Comment (petemounce) [reply defaults/main.yml]: Yup!

File: defaults/main.yml
Comment (petemounce) [reply defaults/main.yml]: Yes, now that I read that doc. Thanks.

File: defaults/main.yml
Comment (petemounce) [reply defaults/main.yml]: So, I shouldn't bother with the v2 template at all?

File: defaults/main.yml
Comment (petemounce) [reply defaults/main.yml]: Ah, it was also only in this branch because of a bad merge on my part. Apologies.

File: defaults/main.yml
Comment (hectcastro) [reply defaults/main.yml]: Na, feel free to focus on only v3 support.

File: defaults/main.yml
Comment (hectcastro) [reply defaults/main.yml:16]: Seems fine to keep then.

File: templates/buildkite-agent.cfg.v3.j2
Comment (petemounce) [inline templates/buildkite-agent.cfg.v3.j2]: Don't merge yet, I'm having a bit of templating pain here.

Comment (petemounce) [reply templates/buildkite-agent.cfg.v3.j2]: Closest I've got to this working is

```
{% set queue = 'queue=' + agent_queue %}

tags={{ queue + buildkite_agent_tags | join(',') }
```

but if I feed tags in as `["test1=foo", "test2=bar"]` I end up with

```
tags=queue=defaulttest1=foo,test2=bar
```

I'm sufficiently unfamiliar with jinja2 templating that this seems harder than it should be...?

**Comment by petemounce** [comment]
I've run this end-to-end and it has worked.

I've got some https://goss.rocks tests that also run and pass - want those in a separate PR?

Also, I'll be doing a bit more ansible, myself, and so likely to want to learn some molecule. Since I think my dependency on buildkite is going to be here to stay, would you welcome some molecule-based tests for this role? Again, separate PR. For this, though, no commitment is being made about timeframe or quality ;)

File: defaults/main.yml
Comment (petemounce) [reply defaults/main.yml:16]: :+1:

File: defaults/main.yml
Comment (hectcastro) [reply defaults/main.yml]: Were you going to remove the v2 agent stuff out in this PR?

File: templates/buildkite-agent.cfg.v3.j2
Comment (hectcastro) [inline templates/buildkite-agent.cfg.v3.j2]: This is likely meant to reference `buildkite_agent_queue`  vs. `agent_queue`.

File: templates/buildkite-agent.cfg.v3.j2
Comment (hectcastro) [inline templates/buildkite-agent.cfg.v3.j2]: I was able to get (what I think is) the desired effect with:

```jinja2
tags={{ ([['queue', buildkite_agent_queue] | join('=')] + buildkite_agent_tags) | join(',') }}
```

File: tasks/main.yml
Comment (hectcastro) [inline tasks/main.yml]: In the meta for this role, it is defined as supporting Ubuntu 14.04 (Trusty). The `systemd` specific requirements in your PR make it so that it doesn't work on 14.04, but potentially does make it work on 16.04.

Can you please replace `trusty` in meta with `xenial`? Also, can you please update the Vagrant box reference in the repository `Vagrantfile` to use:

```ruby
config.vm.box = "bento/ubuntu-16.04"
```

File: tasks/main.yml
Comment (hectcastro) [inline tasks/main.yml]: This file seems to be included in the BK agent package. Also, `copy` without `remote_src: True` leads to the copying of `src` from the local file system to the remote.

I think that we can remove this entire task.

File: tasks/main.yml
Comment (hectcastro) [inline tasks/main.yml]: The `item` variable here does not seem to render. Maybe just make "instance" plural?

**Review by hectcastro — COMMENTED** [review]
Molecule tests would be appreciated. We made an effort to add that to some of the roles we support, but not all.

File: templates/buildkite-agent.cfg.v3.j2
Comment (petemounce) [reply templates/buildkite-agent.cfg.v3.j2]: Yes! Good spot. I think my playbook variable cascaded through and that's why I missed this.

File: tasks/main.yml
Comment (petemounce) [reply tasks/main.yml]: :+1:

File: tasks/main.yml
Comment (petemounce) [reply tasks/main.yml]: Huh, I hadn't noticed that, and just blithely followed BK's docs for doing this. Ok; removed.

(More context - I'm using ansible-local packer provisioner, not the remote, so didn't spot the `src`/`remote_src` bit either).

File: tasks/main.yml
Comment (petemounce) [reply tasks/main.yml]: Hah, TIL. Thanks. Yes, updated.

**Comment by petemounce** [comment]
I saw a question about whether I plan to delete v2. I hadn't planned to - don't know who's using this and didn't want to break them if they're just pulling master via ansible-galaxy and not pinning. I thought I'd leave deprecation as something for you, the author :)

**Review by hectcastro — APPROVED** [review]
Looks good. Can you please attempt to squash down your commits a bit? Afterwards, I'll merge and work on cutting a new release.

**Comment by petemounce** [comment]
Happy to squash commits, but please provide a steer? Would you like a single one (in which case, that's an option on the merge button), or would you like something different? If the latter, I'm not sure what you'd like yet.

**Comment by petemounce** [comment]
Thanks again for the good contribution experience and code review.

**Comment by hectcastro** [comment]
👍 Thanks for taking the time to make the contribution!

**Pull request merged**
